agent/agent.go

```diff
@@ -2,229 +2,169 @@ package agent
 		pubsub.WithMessageSignaturePolicy(pubsub.StrictNoSign),
-		//pubsub.WithNoAuthor(),
-		pubsub.WithPeerOutboundQueueSize(pubsubQueueSize),
-		pubsub.WithValidateQueueSize(pubsubQueueSize),
-		pubsub.WithGossipSubParams(pubsubGossipParam()),
+		pubsub.WithPeerOutboundQueueSize(a.Config.PubsubQueueSize),
+		pubsub.WithValidateQueueSize(a.Config.PubsubQueueSize),
+		pubsub.WithGossipSubParams(a.getPubsubGossipParams()),
 	}
 	ps, err := pubsub.NewGossipSub(context.Background(), host, psOpts...)
 	if err != nil {
-		return nil, err
-	}
-
-	a := &Agent{
-		logger:    logger,
-		host:      host,
-		gossipSub: ps,
-		config:    config,
+		return err
 	}
 
-	if config.HttpAddr != nil {
-		a.setupHttp()
-	}
-
-	// read latency
-	latency := readLatency()
-
-	// try to decode the city
-	if config.City == "" {
-		logger.Printf("Generate new city...")
-		config.City = latency.getRandomCity()
-	}
-
-	// start the proxy
-	logger.Printf("[INFO]: Address: %s", config.Addr.String())
-
-	logrus.SetFormatter(&logrus.JSONFormatter{})
-
-	// Output to stdout instead of the default stderr
-	// Can be any io.Writer, see below for File example
-	logrus.SetOutput(os.Stdout)
-
-	// Only log the warning severity or above.
-	logrus.SetLevel(logrus.WarnLevel)
-
-	// In 5% of the cases not sure why but this port is already in use, then, remove the proxy
-	// and use the normal connection.
-	// I think this should work fine
-
-	// TODO: handle latency here
-
-	logger.Printf("Agent started: addr=%s, city=%s", listenAddr.String(), config.City)
-
-	if config.MaxPeers == 0 {
-		panic("Max peers cannot be zero")
+	// topic
+	topic, err := ps.Join(topicName)
+	if err != nil {
+		return err
 	}
-	if config.MaxPeers > -1 {
-		logger.Printf("Max peers: %d", config.MaxPeers)
+	sub, err := topic.Subscribe()
+	if err != nil {
+		return err
 	}
 
-	peerChan := initMDNS(host, config.RendezvousString)
-	go func() {
-		for {
-			peer := <-peerChan
-			logger.Printf("[INFO] Found peer: peer=%s", peer)
-
-			numPeers := len(host.Network().Peers())
-			if config.MaxPeers != -1 {
-				if numPeers > int(config.MaxPeers) {
-					logger.Printf("[INFO]: Skip peer")
-					continue
-				}
-			}
+	readLoop(sub, host.ID(), a.Config.MsgReceivedFn)
 
-			if err := host.Connect(context.Background(), peer); err != nil {
-				fmt.Println("Connection failed:", err)
-			}
-		}
-	}()
+	a.Host, a.Topic = host, topic
+	return nil
+}
 
-	if a.topic, err = ps.Join("topic"); err != nil {
-		return nil, err
-	}
-	sub, err := a.topic.Subscribe()
+func (a *Agent) Connect(remote network.ClusterAgent) error {
+	remoteAddr := remote.(*Agent).Addr()
+	peer, err := peer.AddrInfoFromP2pAddr(remoteAddr)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	readLoop(sub, func(msg *Msg) {
-		elapsed := time.Since(msg.Time)
-
-		logger.Printf("=> '%s' '%d' '%s' '%s' '%s' '%s'", elapsed, len(host.Network().Peers()), time.Now(), msg.From, msg.Hash, msg.Time)
-	})
+	return a.Host.Connect(context.Background(), *peer)
```

Comment:
How async is this function? I am not sure If I remember correctly from the `libp2p` integration but I think this was async. Could we use the same pattern with `ConnectedF` and channels to make it async? 
Otherwise during topology definition stage we will not be sure how many are really connected at the time we start the gossip.

Reply:
Fixed